### PR TITLE
Minimum throughput body timeouts Pt.1

### PR DESCRIPTION
**Pull request by Velfi**
## Motivation and Context

https://github.com/awslabs/smithy-rs/issues/1562

## Description

This change adds a new body wrapper: The minimum throughput limit wrapper. It tracks the rate that data is being streamed from itself. If that rate falls below some configurable limit, it emits an error instead of the next chunk. This protects users from requests that start quickly but then slow down considerably.

I'd like to get this merged and then figure out the codegen/docs/examples/config part in a separate PR.

## Testing

Tests are included.

----

_By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice._

### CODE REVIEW

File: aws/sdk/integration-tests/s3/tests/throughput-timeout.rs
Comment (Velfi) [inline aws/sdk/integration-tests/s3/tests/throughput-timeout.rs:94]: I haven't yet figure out how to retain access to the `sleep_impl` and `time_source` such that I can run these asserts.

Comment (jdisanti) [reply aws/sdk/integration-tests/s3/tests/throughput-timeout.rs:94]: Should be able to clone them before calling `into_shared()`.

File: aws/sdk/integration-tests/s3/tests/throughput-timeout.rs
Comment (Velfi) [inline aws/sdk/integration-tests/s3/tests/throughput-timeout.rs:155]: This is currently failing due to a hyper error that reminds me of when I was doing the concurrency tests.

```
dispatch failure: io error: error writing a body to connection: Connection reset by peer (os error 54): Connection reset by peer (os error 54) (DispatchFailure(DispatchFailure { source: ConnectorError { kind: Io, source: hyper::Error(BodyWrite, Os { code: 54, kind: ConnectionReset, message: "Connection reset by peer" }), connection: Unknown } }))
```

It's like it's dying when the inner body returns an error. I want to figure this out before merging this PR.

File: aws/sdk/integration-tests/s3/tests/throughput-timeout.rs
Comment (jdisanti) [inline aws/sdk/integration-tests/s3/tests/throughput-timeout.rs:63]: Both panic? What is the passing path?

File: aws/sdk/integration-tests/s3/tests/throughput-timeout.rs
Comment (jdisanti) [inline aws/sdk/integration-tests/s3/tests/throughput-timeout.rs:29]: Seems like these tests should be in minimum_throughput.rs. Doesn't look like they have anything to do with S3.

File: aws/sdk/integration-tests/s3/tests/throughput-timeout.rs
Comment (jdisanti) [inline aws/sdk/integration-tests/s3/tests/throughput-timeout.rs:48]: Make this a function to reduce duplication.

File: rust-runtime/aws-smithy-runtime/src/client/http/body/minimum_throughput.rs
Comment (jdisanti) [inline rust-runtime/aws-smithy-runtime/src/client/http/body/minimum_throughput.rs:34]: How large does this queue get in practice?

File: rust-runtime/aws-smithy-runtime/src/client/http/body/minimum_throughput.rs
Comment (jdisanti) [inline rust-runtime/aws-smithy-runtime/src/client/http/body/minimum_throughput.rs:114]: Seems like this task shouldn't get spawned if we're going to return `Poll::ready`. Maybe split out most of this function into a separate function, and then here, call that function and examine the poll result before spawning this task.

File: rust-runtime/aws-smithy-runtime/src/client/http/body/minimum_throughput.rs
Comment (jdisanti) [inline rust-runtime/aws-smithy-runtime/src/client/http/body/minimum_throughput.rs:129]: Seems like it should pop anything that's older than our time window, with the window size being a large upper bound. 16 is probably not high enough? Need to experiment with that number.

File: aws/sdk/integration-tests/s3/tests/throughput-timeout.rs
Comment (Velfi) [reply aws/sdk/integration-tests/s3/tests/throughput-timeout.rs:63]: I replaced this with something that should make more sense.

File: rust-runtime/aws-smithy-runtime/src/client/http/body/minimum_throughput.rs
Comment (Velfi) [reply rust-runtime/aws-smithy-runtime/src/client/http/body/minimum_throughput.rs]: On my machine, a 100MB download with a tick rate of 1s produces 6700 logs. The download completes in only a few secs.

File: rust-runtime/aws-smithy-async/src/rt/sleep.rs
Comment (jdisanti) [inline rust-runtime/aws-smithy-async/src/rt/sleep.rs:86]: Futures shouldn't be `Sync`.

**Comment by Velfi** [comment]
This PR is currently blocked on some HTTP Body changes that @ysaito1001 will soon PR.

**Comment by Velfi** [comment]
The current CI failures are slightly related to Yuki's work to feature gate stuff, I'll resolve them once he gets that work merged.

File: rust-runtime/aws-smithy-async/src/rt/sleep.rs
Comment (Velfi) [reply rust-runtime/aws-smithy-async/src/rt/sleep.rs:86]: After talking with Carl, I decided that this future should in fact be sync. While we could wrap this future in a mutex or similar to make it `Sync`, I'm opting to just add the bound. The reasoning is that omitting the `Sync` bound will cause more user pain than including it.

Comment (ysaito1001) [reply rust-runtime/aws-smithy-async/src/rt/sleep.rs:86]: Requiring a future to be `Sync` can cause some issue when used with an async block ([lang forum](https://users.rust-lang.org/t/future-is-not-sync-when-calling-async-fn-of-dyn-trait-obj-in-another-async-fn/83285)), but I don't necessarily see how that bites us here 🤔

At least, I think this PR needs a label `breaking-change` because `Sleep::new` now fails to take an async block for certain cases ([playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=59ab7e7a4f8aead7788484a1b1fb339c) - derived from the above lang forum).

File: rust-runtime/aws-smithy-runtime/src/client/http/body/minimum_throughput/throughput.rs
Comment (rcoh) [inline rust-runtime/aws-smithy-runtime/src/client/http/body/minimum_throughput/throughput.rs:73]: I think we can compare `bytes_per_second` in all cases right? That's already been normalized and compares `per_time_elapsed`

File: rust-runtime/aws-smithy-runtime/src/client/http/body/minimum_throughput/throughput.rs
Comment (rcoh) [inline rust-runtime/aws-smithy-runtime/src/client/http/body/minimum_throughput/throughput.rs:122]: what's floating_back?

File: rust-runtime/aws-smithy-runtime/src/client/http/body/minimum_throughput/throughput.rs
Comment (rcoh) [inline rust-runtime/aws-smithy-runtime/src/client/http/body/minimum_throughput/throughput.rs:90]: same thing here—bytes_per_second is already normalized

File: rust-runtime/aws-smithy-runtime/src/client/http/body/minimum_throughput.rs
Comment (rcoh) [inline rust-runtime/aws-smithy-runtime/src/client/http/body/minimum_throughput.rs:130]: is it worth benchmarking this at some point? if we're getting a ton of data how much time do we spend calculating throughput? I wonder if we should _only_ calculate it when the sleep expires, e.g.

File: rust-runtime/aws-smithy-runtime/src/client/http/body/minimum_throughput.rs
Comment (rcoh) [inline rust-runtime/aws-smithy-runtime/src/client/http/body/minimum_throughput.rs:170]: this can be simplified—poll_trailers can only be called once. The only thing we need to do here is watch for a timeout, we don't actually need to calculate the size of the trailers. I wonder if we actually need _anything_ in this method at all—once the last read of the body is complete, we essentially want to schedule a timeout for how long it's OK to wait for the trailers to be polled.

File: rust-runtime/aws-smithy-runtime/src/client/http/body/minimum_throughput.rs
Comment (rcoh) [inline rust-runtime/aws-smithy-runtime/src/client/http/body/minimum_throughput.rs:221]: it isn't really though—we're using SystemTime and not `Instant`. SystemTime doesn't guarantee monotonicity:
> Distinct from the Instant type, this time measurement is not monotonic. This means that you can save a file to the file system, then save another file to the file system, and the second file has a SystemTime measurement earlier than the first. In other words, an operation that happens after another operation in real time may have an earlier SystemTime!

https://doc.rust-lang.org/std/time/struct.SystemTime.html

We should probably try to handle this gracefully or add some sort of `InstantTimeSource` in our bag of tricks

File: rust-runtime/aws-smithy-runtime/src/client/http/body/minimum_throughput.rs
Comment (rcoh) [inline rust-runtime/aws-smithy-runtime/src/client/http/body/minimum_throughput.rs:230]: do any tests attempt to test the trailer reading?

**Review by rcoh — COMMENTED** [review]
Looks good—a few questions

File: rust-runtime/aws-smithy-runtime/src/client/http/body/minimum_throughput.rs
Comment (Velfi) [reply rust-runtime/aws-smithy-runtime/src/client/http/body/minimum_throughput.rs]: I'm delegating the trailer method so I won't test this.

File: rust-runtime/aws-smithy-runtime/src/client/http/body/minimum_throughput/throughput.rs
Comment (Velfi) [reply rust-runtime/aws-smithy-runtime/src/client/http/body/minimum_throughput/throughput.rs:122]: ignore it, I'm removing it in favor of a simpler strategy.

File: rust-runtime/aws-smithy-runtime/src/client/http/body/minimum_throughput.rs
Comment (Velfi) [reply rust-runtime/aws-smithy-runtime/src/client/http/body/minimum_throughput.rs]: Agreed, just not in this PR.

File: rust-runtime/aws-smithy-runtime/src/client/http/body/minimum_throughput.rs
Comment (Velfi) [reply rust-runtime/aws-smithy-runtime/src/client/http/body/minimum_throughput.rs]: Agreed, just not in this PR.

**Review by rcoh — APPROVED** [review]

**Pull request merged**
